src/compiler-node.js

```diff
@@ -1,111 +1,61 @@
-    source = source.replace(/^\.\//, '');
-    target = target.replace(/^\.\//, '');
-
-    walker.on("file", function(root, fileStats, next) {
-      if (fileStats.name[0] == ".") {
-          return next();
-      }
-
-      if (path.extname(fileStats.name) !== '.bs') {
-          return next();
-      }
-
-      var sourcePath = path.join(root, fileStats.name);
-      var targetRoot = root.replace(source, target);
-      var targetPath = sourcePath.replace(source, target).replace('.bs', '.js');
-
-      mkdir.sync(targetRoot);
-      parseFile(root, sourcePath, targetPath, options, onFile, onError, next);
-    });
-
-    walker.on("end", function () {
-      if (onDone) onDone();
-    });
+function parseFiles (source, target, options, callback) {
+    return mkdirp(target)
+        .then(function() {
+            return Bluebird.all([fs.lstatAsync(source), fs.lstatAsync(target)]);
+        })
+        .spread(function (sourceStat, targetStat) {
+            source = utils.checkDir(source, sourceStat);
+            target = utils.checkDir(target, targetStat);
+
+            // If it's only one file, we compile directly
+            if (!sourceStat.isDirectory()) {
+                // bailey bs/main.bs build/ should put main.js in build/
+                if (targetStat.isDirectory()) {
+                    target += path.basename(source);
+                }
+                return parseFile(source, target, options);
+            }
+
+            // From here on we need a ./ from the start to be removed
+            source = source.replace(/^\.\//, '');
+            target = target.replace(/^\.\//, '');
+
+            return utils.readDirAsync('./' + source, utils.fileFilter)
+                .each(function(sourcePath) {
+                    var targetPath = sourcePath.replace(source, target).replace('.bs', '.js');
+                    return parseFile(sourcePath, targetPath, options);
+                });
+            }).nodeify(callback);
 }
 
-function parseFile (root, sourcePath, targetPath, options, onFile, onError, onDone) {
-  var startTime = Date.now();
-
-  fs.readFile(sourcePath, 'utf8', function (err, input) {
-
-    if (err) {
-      if (onError) onError(err);
-      return;
-    }
-
-    options.filePath = sourcePath;
-    options.root = root;
-
-    var parsed;
-
-    try {
-      parsed = compiler.parse(input, options, onError);
-    }
-    catch (e) {
-      if (onError) onError(e);
-    }
-
-    if (parsed !== undefined) {
-      fs.writeFile(targetPath, parsed, function(err) {
-        if (err) {
-          if (onError) onError(err);
-          return;
-        }
-        if (onFile) onFile(sourcePath, targetPath, Date.now() - startTime);
-        if (onDone) onDone();
-      });
-
-    }
-
-  });
+function parseFile (sourcePath, targetPath, options, callback) {
+    return fs.readFileAsync(sourcePath, 'utf8')
+        .then(function(input) {
+            return new Bluebird(function(resolve, reject) {
```

Comment:
Could do something like this instead of using `new Bluebird`:
```js
        .then(function(input) {
            var parsed = compiler.parse(input, options);
            if (parsed === undefined) throw AmazingError();
            return fs.writeFileAsync(targetPath, parsed);
        })
```